### DISCUSSION TITLE
Why does it crash?

### DISCUSSION THREAD

**Discussion by throttlemeister** [abraunegg/onedrive#1788] [Q&A] [answered]
I am finding onedrive terminates with a SIGPIPE command every few minutes on Fedora (didn't notice a problem on Pop! (Ubuntu-based)). I can change the systemd script to set the restart policy to always, but that's not really a satisfactory solution.

Found several references also in the bugtracker, but the overal sentiment seems to be its not a problem with the onedrive app, but a network issue. From the reports, there does not appear to be a relation to a specific OS or version.

I have a problem with this explanation, because:
1) When I was using onedrive on Pop!_OS, I did not experience this issue on the same computers. Note plural: both my laptop and desktop were working fine. Now that I run Fedora on both, onedrive terminates after whatever time period is set for monitor on both. At least at first glance, there appears to be a relationship between onedrive and the OS in this.
2) If I install a proxy server on my NAS, and tell onedrive to use a proxy server for http and https, it does not terminate anymore. Ever. If there was a network issue, it would not be solved by using a proxy in between

Thoughts?

**Comment by abraunegg** [comment]
@throttlemeister 
> Found several references also in the bugtracker, but the overal sentiment seems to be its not a problem with the onedrive app, but a network issue. From the reports, there does not appear to be a relation to a specific OS or version

The issue 99% of the time has been seen in the following circumstances:
1. Users using Ubuntu (not Debian, not others). It is something in the Ubuntu platform that significantly contributes to this.
2. Users in Europe who are connecting through to Microsoft Eastern European Data Center

The remaining times this has popped up it has been:
1. Bad network connectivity - wifi or other
2. Bad DNS queries

> I have a problem with this explanation, because
> ...

Actually this just confirms that it is something in the Linux OS that is significantly contributing to this issue.

When Ubuntu users, who faced the issue changed their OS or tested with something other than Ubuntu, the issue, on the same hardware, same network *went away* - confirming that that this is not an application issue but something related to the Operating System | Linux Kernel | TCP stack. When Ubuntu users also disabled their *ufw* - the issue manifestation also decreased.

Your experience is the reverse - did not experience the issue on your prior OS, but on Fedora you do .. however it basically confirms that this is a Linux OS related issue.

> If I install a proxy server on my NAS, and tell onedrive to use a proxy server for http and https, it does not terminate anymore. Ever. If there was a network issue, it would not be solved by using a proxy in between

Right .. because your proxy server is now handing the TCP states differently to the client which is using 'curl' - so the interaction from the host OS to the proxy server (even if you have say squid installed locally) is changing the TCP handling from the 'client' to OneDrive. I suspect the proxy server, being on a NAS is also a totally separate system to your system where you are running OneDrive from - thus the TCP connection from NAS -> OneDrive is being handled there - thus, all the *other* interesting stuff is being handled locally. I wonder if your NAS / proxy server is showing any *high* rate of client errors / connection issues | failures and is gracefully just accepting / reconnecting things for the client, masking the current issue ..

How the client works is:
* client creates OneDrive API classes -> calls your 'curl' library on your Linux System. This is why when a bug report is opened one of the items being asked for is the 'curl' version .. because, in the past, when 'curl' have changed options to be default (such as defaulting to HTTP2 which OneDrive service does not support) - can cause issues.
* 'curl' is what makes the call to OneDrive

When the application stops without reason - it is because 'curl' has failed / exited uncleanly - hence the 141 error as the exit code. From all debugging that has been done (and and cannot stress enough I have never ever been able to replicate this issue myself) - the way it manifests to me feels like that the host operating system is the one that is mis-handling the TCP connection state, and closing the connection.

So .. I would be looking at the differences in the following areas:

1. What 'curl' version is installed on Fedora and your pop_os platform - are there differences there - what options has it been compiled with?
2. What are the TCP options deployed | used at the kernel level on both systems - systemctl options around IPv4 TCP keepalive - `/proc/sys/net/ipv4/tcp_keepalive_time` - is there a difference - is there something obvious?
3. What are the actual 'kernel' TCP compiled in differences - getting the kernel .config file and comparing what is set TCP wise may garnish some clues

**Comment by abraunegg** [comment]
@throttlemeister 
Some follow up questions for you:
1. What version of Fedora are you seeing this constantly on?
2. What version of Pop!_OS did you *not* experience this on?
3. What version of the application were you using on each OS?
4. Did you compile from source | use package on Fedora?
5. If you installed from a package, what happens if compile from source on Fedora (after ensuring that no package / onedrive binary exists on your system) - do you get the same issue?

**Reply by throttlemeister** [reply]
1) F35
2) 21.10
3) Obs version as per documentation for pop specific version don't know(no longer installed, max 2 months ago), fedora package 2.4.14
4) not tried from source yet
5) need to get back on that.

**Reply by throttlemeister** [reply]
Removed the Fedora package and installed from source.

Version: onedrive v2.4.15-4-g6f53580

Result: process dies on a SIGPIPE after 2.5 minutes when triggering sync activity by removing a file from the local filesystem in the OneDrive directory.

tcp_keep_alive_time = 7200

curl 7.79.1 (x86_64-redhat-linux-gnu) libcurl/7.79.1 OpenSSL/1.1.1l-fips zlib/1.2.11 brotli/1.0.9 libidn2/2.3.2 libpsl/0.21.1 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.45.1 OpenLDAP/2.4.59

**Reply by abraunegg** [reply]
@throttlemeister 
Thanks - currently installing Fedora 35 to see if I can replicate this on Fedora.

Does the SIGPIPE exit only occur when using a systemd service or also when running manually:
```
onedrive --monitor --verbose <plus any other options needed for your env>
```

**Reply by throttlemeister** [reply]
Yes. Started onedrive --monitor --verbose and it read my config and processed and updated a sync. Once it was finished, I let it sit for about 2 minutes then updated a file in OneDrive and it immediately terminated without even as much as a message.

Sync with OneDrive is complete
[M] Item deleted: Documents/testfile
Deleting item from OneDrive: Documents/testfile
╭─(throttlemeister웃loki.sport-touring.eu)-[~/OneDrive/Documents] 
╰─$ 

Note the file is not yet removed from OneDrive as the sync hasn't happened yet. If I now restart OneDrive, it will sync and do the actual delete.

**Comment by abraunegg** [comment, marked as answer]
@throttlemeister 
Whilst I am still waiting for Fedora 35 to install to try and replicate your issue - please can you test this PR:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/1789/head:pr1789
git checkout pr1789
./configure; make clean; make;
```
To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

When running the PR, your version should be: `onedrive v2.4.15-5-g9d986fa` or greater.

**Reply by throttlemeister** [reply]
Running this PR version appears to have cured the sigpipe crashing, at least for the last 5+ minutes of testing so far. It has processed changes made to the directory as well as processing the scheduled update after 5.

I'll keep it running for a while, but it looks promising.

**Reply by abraunegg** [reply]
@throttlemeister 
OK - thanks ... very interesting ... all the PR is doing is setting the options for a few items to their supposed 'defaults' - so it is 'odd' that these are being set to something other than the supposed default somewhere ....

**Reply by throttlemeister** [reply]
> @throttlemeister OK - thanks ... very interesting ... all the PR is doing is setting the options for a few items to their supposed 'defaults' - so it is 'odd' that these are being set to something other than the supposed default somewhere ....

Interesting indeed. I have installed this version, and it's still running and syncing. So looks good. But as to why the defaults are not default, very strange indeed. I don't even have a config file as I don't have a reason to deviate from the defaults.

**Reply by abraunegg** [reply]
@throttlemeister 
By 'defaults' here I mean the libcurl defaults as the application calls libcurl .... reading the build logs of curl on koji (fedoras build platorm) (https://kojipkgs.fedoraproject.org//packages/curl/7.79.1/1.fc35/data/logs/x86_64/build.log) nothing stands out that something is being set / changed

**Comment by abraunegg** [comment]
@throttlemeister 
FYI - currently running Fedora 35 + 'onedrive' package version 2.4.14 .. and as of now, unable to reproduce this SIGPIPE issue at all .. so this is also potentially something environmental as a contributing cause .. something setup | configured | installed on Fedora beyond the default install ?

**Reply by throttlemeister** [reply]
Well, its a running, working system i use to do work on so yeah, it's not vanilla anymore. But it's not like there is all kinds of exotic stuff on here, just some basic tools and office programs. Basically either rpms, flatpak or appimage. Nothing build from scratch that may have installed some libraries, other than now onedrive.

**Reply by abraunegg** [reply]
@throttlemeister 
Do any of those 'tools' or flatpak / appimage installs require or use curl or libcurl as dependency ?

**Reply by throttlemeister** [reply]
@abraunegg negative. Flatpak and AppImage are both containerized applications anyway, so it would be very strange and very wrong if they needed something on the system outside of their containerized environment.

**Reply by throttlemeister** [reply]
The only non-standard packages (ie non-Fedora repos) installed are MS Edge & vscode, nvidia drivers and multimedia codecs / programs from rpmfusion. Everything else comes from the Fedora repositories.

**Reply by abraunegg** [reply]
@throttlemeister 
OK .. so really it comes down to libcurl on Fedora then .... but also I cannot replicate at all - running here without issue for the last 20 mins

**Comment by abraunegg** [comment]
@throttlemeister 
As a comparison:

**POP_OS! 21.10 Curl Version**
```
curl 7.74.0 (x86_64-pc-linux-gnu) libcurl/7.74.0 OpenSSL/1.1.1l zlib/1.2.11 brotli/1.0.9 libidn2/2.3.1 libpsl/0.21.0 (+libidn2/2.3.0) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3
Release-Date: 2020-12-09
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

**Fedora 35 Curl Version**
```
curl 7.79.1 (x86_64-redhat-linux-gnu) libcurl/7.79.1 OpenSSL/1.1.1l-fips zlib/1.2.11 brotli/1.0.9 libidn2/2.3.2 libpsl/0.21.1 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.45.1 OpenLDAP/2.4.59
Release-Date: 2021-09-22
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

There are some subtle differences here between the compiled options / feature support and what libraries are being used to compile 'curl'

Going through https://curl.se/changes.html to look at all the code changes is not yielding much information regarding SIGPIPE - there are a few references between those version but nothing that stands out as obvious, and Debian|Ubuntu|POP developers might also be applying backported patches or maybe even tweaking something else in-between.

**Reply by throttlemeister** [reply]
Basically the only difference is HSTS support in Fedora. Weird. Generally specifying default values for a supporting program isn't a problem I guess? Curious as to if this fixes the problems for other people that had this issue as well.

**Reply by abraunegg** [reply]
@throttlemeister 
In all the other reported issues over the years - all we had was the exit code - nothing as per your experience / additional information that provided some further insight as to SIGPIPE.

If you can keep running the PR version for a while - 24/48 hrs, and report back - if all good I will merge that PR into 'master'

**Reply by throttlemeister** [reply]
@abraunegg I will let you know tomorrow or day after how it goes. Right now, still happily purring along without crash or issue.

**Reply by throttlemeister** [reply]
@abraunegg still going strong, also after a suspend period. No errors, no crashes.

**Reply by abraunegg** [reply]
@throttlemeister 
Thanks for the feedback - will merge the PR into 'master'
